File: app/src/main/java/org/ajar/swadesh/model/ipa/Vowel.java
package org.ajar.swadesh.model.ipa;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

import static org.ajar.swadesh.model.ipa.Backness.*;
import static org.ajar.swadesh.model.ipa.Openness.*;
import static org.ajar.swadesh.model.ipa.Rounding.*;
import static org.ajar.swadesh.model.ipa.RColor.*;

public enum Vowel implements PhoneticSymbol<Vowel> {
    i("i", CLOSE, FRONT, UNROUNDED),
    ĩ("ĩ", CLOSE, FRONT, UNROUNDED, NASALIZED),
    y("y", CLOSE, FRONT, ROUNDED),
    ɨ("ɨ", CLOSE, CENTRAL, UNROUNDED),
    ʉ("ʉ", CLOSE, CENTRAL, ROUNDED),
    ɯ("ɯ", CLOSE, BACK, UNROUNDED),
    u("u", CLOSE, BACK, ROUNDED),
    ũ("ũ", CLOSE, BACK, ROUNDED, NASALIZED),

    ɪ("ɪ", NEAR_CLOSE, FRONT, UNROUNDED),
    ʏ("ʏ", NEAR_CLOSE, FRONT, ROUNDED),
    ɨ_("ɨ", NEAR_CLOSE, CENTRAL, UNROUNDED),
    ʉ̞_("̞ʉ̞", NEAR_CLOSE, CENTRAL, ROUNDED),
    ɯ̞("ɯ̞", NEAR_CLOSE, BACK, UNROUNDED),
    ʊ("ʊ", NEAR_CLOSE, BACK, ROUNDED),

    e("e", CLOSE_MID, FRONT, UNROUNDED),
    ẽ("ẽ", CLOSE_MID, FRONT, UNROUNDED, NASALIZED),
    ø("ø", CLOSE_MID, FRONT, ROUNDED),
    ɘ("ɘ", CLOSE_MID, CENTRAL, UNROUNDED),
    ɵ("ɵ", CLOSE_MID, CENTRAL, ROUNDED),
    ɤ("ɤ", CLOSE_MID, BACK, UNROUNDED),
    o("o", CLOSE_MID, BACK, ROUNDED),
    õ("õ", CLOSE_MID, BACK, ROUNDED, NASALIZED),

    e̞("e̞", MID, FRONT, UNROUNDED),
    ø̞("ø̞", MID, FRONT, ROUNDED),
    ə("ə", MID, CENTRAL, UNROUNDED),
    ɚ("ɚ", MID, CENTRAL, UNROUNDED, RHOTIC),
    ɤ̞("ɤ̞", MID, BACK, UNROUNDED),
    o̞("o̞", MID, BACK, ROUNDED),

    ɛ("ɛ", OPEN_MID, FRONT, UNROUNDED),
    œ("œ", OPEN_MID, FRONT, ROUNDED),
    ɜ("ɜ", OPEN_MID, CENTRAL, UNROUNDED),
    ɝ("ɝ", OPEN_MID, CENTRAL, UNROUNDED, RHOTIC),
    ɞ("ɞ", OPEN_MID, CENTRAL, ROUNDED),
    ʌ("ʌ", OPEN_MID, BACK, UNROUNDED),
    ɔ("ɔ", OPEN_MID, BACK, ROUNDED),

    æ("æ", NEAR_OPEN, FRONT, UNROUNDED),
    ɐ("ɐ", NEAR_OPEN, CENTRAL, UNROUNDED),

    a("a", OPEN, FRONT, UNROUNDED),
    ã("ã", OPEN, FRONT, UNROUNDED, NASALIZED),
    ɶ("ɶ", OPEN, FRONT, ROUNDED),
    ä("ä", OPEN, CENTRAL, UNROUNDED),
    ɒ̈("ɒ̈", OPEN, CENTRAL, ROUNDED),
    ɑ("ɑ", OPEN, BACK, UNROUNDED),
    ɑ_n("̃d", OPEN, BACK, UNROUNDED, NASALIZED),
    ɒ("ɒ", OPEN, BACK, ROUNDED);

    private final static List<Vowel> searchList = new ArrayList<>();

    private final String symbol;
    private final Rounding round;
    private final Openness open;
    private final Backness back;
    private final RColor color;

    Vowel(String symbol, Openness open, Backness back, Rounding round) {
        this(symbol, open, back, round, RColor.NOT_RHOTIC);
    }

    Vowel(String symbol, Openness open, Backness back, Rounding round, RColor color) {
        this.symbol = symbol;
        this.round = round;
        this.open = open;
        this.back = back;
        this.color = color;
    }

    @Override
    public int length() {
        return symbol.length();
    }

    @Override
    public String description() {
        return open.toString() + ", " + back.toString() + ", " + round.toString() + " vowel";
    }

    @Override
    public boolean starts(String string) {
        return string.startsWith(symbol);
    }

    @Override
    public double distanceTo(Vowel location) {
        //In theory, there should be more distance between adjacent sounds the more closed you get. That's not implemented here.
        double roundD = round.distanceTo(location.round);
        double openD = open.distanceTo(location.open);
        double backD = back.distanceTo(location.back);
        double colorD = color.distanceTo(location.color);
        return Math.sqrt((roundD * roundD) + (openD * openD) + (backD * backD) + (colorD * colorD));
    }

    public static List<Vowel> getSearchList() {
        if(searchList.size() == 0) {
            searchList.addAll(Arrays.asList(Vowel.values()));
            searchList.sort(new Comparator<Vowel>() {
                @Override
                public int compare(Vowel o1, Vowel o2) {
                    return o2.length() - o1.length();
                }
            });
        }

        return searchList;
    }
}

File: app/src/main/java/org/ajar/swadesh/model/ipa/Vowel.kt
package org.ajar.swadesh.model.ipa

import java.util.*
import kotlin.math.sqrt

enum class Vowel @JvmOverloads constructor(private val symbol: String, private val open: Openness, private val back: Backness, private val round: Rounding, private val color: RColor = RColor.NOT_RHOTIC) : PhoneticSymbol {
    i("i", Openness.CLOSE, Backness.FRONT, Rounding.UNROUNDED),
    ĩ("ĩ", Openness.CLOSE, Backness.FRONT, Rounding.UNROUNDED, RColor.NASALIZED),
    y("y", Openness.CLOSE, Backness.FRONT, Rounding.ROUNDED),
    ɨ("ɨ", Openness.CLOSE, Backness.CENTRAL, Rounding.UNROUNDED),
    ʉ("ʉ", Openness.CLOSE, Backness.CENTRAL, Rounding.ROUNDED),
    ɯ("ɯ", Openness.CLOSE, Backness.BACK, Rounding.UNROUNDED),
    u("u", Openness.CLOSE, Backness.BACK, Rounding.ROUNDED),
    ũ("ũ", Openness.CLOSE, Backness.BACK, Rounding.ROUNDED, RColor.NASALIZED),
    ɪ("ɪ", Openness.NEAR_CLOSE, Backness.FRONT, Rounding.UNROUNDED),
    ʏ("ʏ", Openness.NEAR_CLOSE, Backness.FRONT, Rounding.ROUNDED),
    ɨ_("ɨ", Openness.NEAR_CLOSE, Backness.CENTRAL, Rounding.UNROUNDED),
    ʉ_dot("̞ʉ̞", Openness.NEAR_CLOSE, Backness.CENTRAL, Rounding.ROUNDED),
    ɯ_dot("ɯ̞", Openness.NEAR_CLOSE, Backness.BACK, Rounding.UNROUNDED),
    ʊ("ʊ", Openness.NEAR_CLOSE, Backness.BACK, Rounding.ROUNDED),
    e("e", Openness.CLOSE_MID, Backness.FRONT, Rounding.UNROUNDED),
    ẽ("ẽ", Openness.CLOSE_MID, Backness.FRONT, Rounding.UNROUNDED, RColor.NASALIZED),
    ø("ø", Openness.CLOSE_MID, Backness.FRONT, Rounding.ROUNDED),
    ɘ("ɘ", Openness.CLOSE_MID, Backness.CENTRAL, Rounding.UNROUNDED),
    ɵ("ɵ", Openness.CLOSE_MID, Backness.CENTRAL, Rounding.ROUNDED),
    ɤ("ɤ", Openness.CLOSE_MID, Backness.BACK, Rounding.UNROUNDED),
    o("o", Openness.CLOSE_MID, Backness.BACK, Rounding.ROUNDED),
    õ("õ", Openness.CLOSE_MID, Backness.BACK, Rounding.ROUNDED, RColor.NASALIZED),
    e_dot("e̞", Openness.MID, Backness.FRONT, Rounding.UNROUNDED),
    ø_dot("ø̞", Openness.MID, Backness.FRONT, Rounding.ROUNDED),
    ə("ə", Openness.MID, Backness.CENTRAL, Rounding.UNROUNDED),
    ɚ("ɚ", Openness.MID, Backness.CENTRAL, Rounding.UNROUNDED, RColor.RHOTIC),
    ɤ_dot("ɤ̞", Openness.MID, Backness.BACK, Rounding.UNROUNDED),
    o_dot("o̞", Openness.MID, Backness.BACK, Rounding.ROUNDED),
    ɛ("ɛ", Openness.OPEN_MID, Backness.FRONT, Rounding.UNROUNDED),
    œ("œ", Openness.OPEN_MID, Backness.FRONT, Rounding.ROUNDED),
    ɜ("ɜ", Openness.OPEN_MID, Backness.CENTRAL, Rounding.UNROUNDED),
    ɝ("ɝ", Openness.OPEN_MID, Backness.CENTRAL, Rounding.UNROUNDED, RColor.RHOTIC),
    ɞ("ɞ", Openness.OPEN_MID, Backness.CENTRAL, Rounding.ROUNDED),
    ʌ("ʌ", Openness.OPEN_MID, Backness.BACK, Rounding.UNROUNDED),
    ɔ("ɔ", Openness.OPEN_MID, Backness.BACK, Rounding.ROUNDED),
    æ("æ", Openness.NEAR_OPEN, Backness.FRONT, Rounding.UNROUNDED),
    ɐ("ɐ", Openness.NEAR_OPEN, Backness.CENTRAL, Rounding.UNROUNDED),
    a("a", Openness.OPEN, Backness.FRONT, Rounding.UNROUNDED),
    ã("ã", Openness.OPEN, Backness.FRONT, Rounding.UNROUNDED, RColor.NASALIZED),
    ɶ("ɶ", Openness.OPEN, Backness.FRONT, Rounding.ROUNDED),
    ä("ä", Openness.OPEN, Backness.CENTRAL, Rounding.UNROUNDED),
    ɒ_dot("ɒ̈", Openness.OPEN, Backness.CENTRAL, Rounding.ROUNDED),
    ɑ("ɑ", Openness.OPEN, Backness.BACK, Rounding.UNROUNDED),
    ɑ_n("̃d", Openness.OPEN, Backness.BACK, Rounding.UNROUNDED, RColor.NASALIZED),
    ɒ("ɒ", Openness.OPEN, Backness.BACK, Rounding.ROUNDED);

    private val components: MutableList<Location>
    override fun length(): Int {
        return symbol.length
    }

    override fun description(): String {
        return "$open, $back, $round vowel"
    }

    override fun starts(string: String): Boolean {
        return string.startsWith(symbol)
    }

    override fun components(): List<Location> {
        return components
    }

    override fun distanceTo(location: Location): Double {
        //In theory, there should be more distance between adjacent sounds the more closed you get. That's not implemented here.
        return when(location) {
            is Vowel -> {
                val roundD = round.distanceTo(location.round)
                val openD = open.distanceTo(location.open)
                val backD = back.distanceTo(location.back)
                val colorD = color.distanceTo(location.color)
                sqrt(roundD * roundD + openD * openD + backD * backD + colorD * colorD)
            }
            else -> -1.0
        }
    }

    companion object {
        private val searchList: MutableList<Vowel?> = ArrayList()
        @JvmStatic
        fun getSearchList(): List<Vowel?> {
            if (searchList.size == 0) {
                searchList.addAll(listOf(*values()))
                searchList.sortWith(Comparator { o1, o2 -> o2!!.length() - o1!!.length() })
            }
            return searchList
        }

        fun find(locations: Collection<Location?>): PhoneticSymbol? {
            return getSearchList().stream().filter { vowel: Vowel? -> vowel!!.components().size == locations.size && vowel.components().containsAll(locations) }.findFirst().orElse(null)
        }
    }

    init {
        components = ArrayList()
        components.add(round)
        components.add(open)
        components.add(back)
        components.add(color)
    }
}